# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0

In [41]:
# Imports section
import pandas as pd
import sklearn as sk
import warnings

# Suppress specific warnings
warnings.filterwarnings("ignore", category=sk.exceptions.ConvergenceWarning)

## 1. Loading the dataset

In [42]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

dataset = pd.read_csv("science_data_large.csv")

print("First 15 Rows:")
print(dataset.head(15))

print("\n Table Information")
print(dataset.info())

First 15 Rows:
    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

 Table Information
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   

## 2. Splitting the dataset

In [43]:
# Take the pandas dataset and split it into our features (X) and label (y)

X = dataset[['Temperature °C', 'Mols KCL']]  # Features should be in a DataFrame
y = dataset['Size nm^3']                    # Labels should be in a Series

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 

X_train, X_test, y_train, y_test = sk.model_selection.train_test_split(X, y, test_size=0.1, random_state=42)


## 3. Perform a Linear Regression

In [44]:
# Use sklearn to train a model on the training set

model = sk.linear_model.LinearRegression()

model.fit(X_train,y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sampleDatapoint = X_test.iloc[0].values.reshape(1,-1)
predicted = model.predict(sampleDatapoint)

score = model.score(X_test,y_test)

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
print("R^2 score")
print(score)
print("\n")

#Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX

# Below is the latex
# y = 866.1464 \cdot x_1 + 1032.6951 \cdot x_2 - 409391.4796
print("Coefficeints followed by intercept")
print(model.coef_)
print(model.intercept_)




R^2 score
0.8552472077276095


Coefficeints followed by intercept
[ 866.14641337 1032.69506649]
-409391.47958340764


/home/codespace/.local/lib/python3.12/site-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Write the linear equation of a slime: (example equation: $E = mc^2$):  $\cdot x_1$ is Temperature and $\cdot x_2$ is Mols KCL

$h(x) = 866.1464 \cdot x_1 + 1032.6951 \cdot x_2 - 409391.4796$

Report on score and explain meaning:
The score in this model represents the R^2 function, with possible values from -inf to 1, with the closer the value to 1 meaning a more accurate regression model. It tells you how accurate the X values are in predicting the y value. Our value of R^2 = 0.85524 is pretty good as extremely high values to 1 could also be a result of overfitting.

## 4. Use Cross Validation

In [45]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
num = sk.model_selection.KFold(n_splits=5,shuffle=True,random_state=42)

scores = sk.model_selection.cross_val_score(model, X, y, cv=num,scoring='r2')

print("R^2 Scores through 5 batches:")
print(scores)
print("Mean of the R^2:")
print(sum(scores)/len(scores))

# Report on their finding and their significance



R^2 Scores through 5 batches:
[0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Mean of the R^2:
0.8597294202684646


Write findings here: The model achieved a mean R^2 score of 0.8597, indicating it explains about 86% of the variance in the target variable, demonstrating strong predictive performance. The scores across the five folds are consistent, suggesting the model generalizes well and is not overly sensitive to different data splits. Overall, this is a good result with no glaring errors.

## 5. Using Polynomial Regression

In [46]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
num = sk.model_selection.KFold(n_splits=5, shuffle=True, random_state=42)

poly = sk.preprocessing.PolynomialFeatures(2)
X_poly = poly.fit_transform(X)

X_train, X_test, y_train, y_test = sk.model_selection.train_test_split(X_poly, y, test_size=0.1, random_state=42)
polymodel = sk.linear_model.LinearRegression()
polymodel.fit(X_train, y_train)


# Perform cross-validation
scores = sk.model_selection.cross_val_score(polymodel, X_poly, y, cv=num, scoring='r2')
    
# Print model coefficients
print("Coefs:")
print(polymodel.coef_)
print("Intercepts")
print(polymodel.intercept_)

# Print Cross Val scores
print("Cross Val Scores:")
print(scores)

# Report on the metrics and output the resultant equation as you did in Part 3.

Coefs:
[ 0.00000000e+00  1.20000000e+01 -1.27195488e-07  1.26494371e-11
  2.00000000e+00  2.85714287e-02]
Intercepts
2.0477105863392353e-05
Cross Val Scores:
[1. 1. 1. 1. 1.]


Polynomial Equation:
$ h(x) = (2.0479 \times 10^{-5}) + (0.0)x + (12.0)x^2 + (-1.27198 \times 10^{-7})x^3 + (1.26473 \times 10^{-11})x^4 + (2.0)x^5 + (0.02857)x^6 $


Report on the score and interpret: The R^2 value is 1.0, predicted in every fold of K-fold cross validation. As previously mentioned, the closer the score is to 1, the better the prediction the model gives. A score of 1.0 means the model likely overfits the data, as every value is correctly predicted in the test set. This is rare in real world applications, but the dataset used for this is synthetic, so this high of a score is expected.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [47]:
# Load the dataset. Then train and evaluate the classification models.

ckd_ds = pd.read_csv("ckd_feature_subset.csv")
#print(ckd_ds.info())
rand = 42
X = ckd_ds[['age', 'bp', 'wbcc', 'appet_poor', 'appet_good','rbcc']] 
y = ckd_ds['Target_ckd']                 


X_train, X_test, y_train, y_test = sk.model_selection.train_test_split(X, y, test_size=0.1, random_state=rand)

# Logistic Regression: 

LR_model = sk.linear_model.LogisticRegression()

LR_model.fit(X_train,y_train)


# SVC

SVC_model = sk.svm.SVC(decision_function_shape='ovo')

SVC_model.fit(X_train,y_train)

# k-Nearest Neighbors

kNear_model = sk.neighbors.KNeighborsClassifier()

kNear_model.fit(X_train,y_train)

# Neural Networks
base_iterations = 1000  # Original 1000
base_activation = 'relu'  # Original ReLU
base_hidden_layers = (100,)  # Original 100 hidden layers

# Configurations to change one variable at a time
configurations = [
    {'iterations': base_iterations, 'activation': base_activation, 'hidden_layers': base_hidden_layers},
    {'iterations': 200, 'activation': base_activation, 'hidden_layers': base_hidden_layers},
    {'iterations': 500, 'activation': base_activation, 'hidden_layers': base_hidden_layers},
    {'iterations': base_iterations, 'activation': 'logistic', 'hidden_layers': base_hidden_layers},
    {'iterations': base_iterations, 'activation': 'identity', 'hidden_layers': base_hidden_layers},
    {'iterations': base_iterations, 'activation': base_activation, 'hidden_layers': (1,)},
    {'iterations': base_iterations, 'activation': base_activation, 'hidden_layers': (500,)},
]

# Iterate through configurations and train models
neurall = []
for config in configurations:
    iterations = config['iterations']
    activationfun = config['activation']
    layer_size = config['hidden_layers']
    
    # Create the neural network model
    neural = sk.neural_network.MLPClassifier(
        max_iter=iterations,
        activation=activationfun,
        hidden_layer_sizes=layer_size
    )
    
    # Train the model
    neurall.append(neural.fit(X_train, y_train))


## Results and Conclusion for Classification Experiments

In [48]:
# Testing for Linear Regression Model: 

cfv = sk.model_selection.cross_val_score(LR_model, X, y, cv=5)
print("%0.2f accuracy with a standard deviation of %0.2f" % (cfv.mean(), cfv.std()))

# Testing for SVC Model

cfv1 = sk.model_selection.cross_val_score(SVC_model, X, y, cv=5)
print("%0.2f accuracy with a standard deviation of %0.2f" % (cfv1.mean(), cfv1.std()))

# Testing for k-Neighbors Model

cfv2 = sk.model_selection.cross_val_score(kNear_model, X, y, cv=5)
print("%0.2f accuracy with a standard deviation of %0.2f" % (cfv2.mean(), cfv2.std()))

# Testing the neural network: 
print("Neural Networks below, first is the base network:")
tt = ["Iterations Changed:","Activation Function Changed:","Hidden Layers Changed:"]
j = 0
for i,v in enumerate(neurall):
    if i != 0 and ((i-1)%2 == 0):
        print(tt[j])
        j+=1

    cfv3 = sk.model_selection.cross_val_score(v, X, y, cv=5)
    print("%0.2f accuracy with a standard deviation of %0.2f" % (cfv3.mean(), cfv3.std()))




0.86 accuracy with a standard deviation of 0.05
0.93 accuracy with a standard deviation of 0.06
0.93 accuracy with a standard deviation of 0.03
Neural Networks below, first is the base network:
0.94 accuracy with a standard deviation of 0.02
Iterations Changed:
0.88 accuracy with a standard deviation of 0.06
0.93 accuracy with a standard deviation of 0.04
Activation Function Changed:
0.92 accuracy with a standard deviation of 0.03
0.94 accuracy with a standard deviation of 0.02
Hidden Layers Changed:
0.71 accuracy with a standard deviation of 0.24
0.95 accuracy with a standard deviation of 0.04


## Results and Conclusion for Neural Network Experiments

| Model                   | Mean Accuracy| Standard Deviation |
|-------------------------|--------------|--------------------|
| **Logistic Regression** | 0.86         | 0.05               |
| **SVM**                 | 0.93         | 0.06               |
| **K-Nearest Neighbors** | 0.93         | 0.03               |
| **Neural Network**      | 0.94         | 0.02               |



I think the best model is the Neural Network, using the base parameters given by scikit other than using 1,000 iterations. This is because it has both the highest mean accuracy and the lowest standard deviaiton, indicating the highest performing results on average with the lowest amount of swing the results will have. 


When training the Neural Network, I changed up three variables inside the MLPClassifier algorithm that ultimalely trains the neural network. These three variables are the following. The training iterations, deciding how many times should this neural network be ran and the weights inside of it changed. The Activation Function, being the function that transforms the function from a linear function to something non-linear, resulting in curves and bends, something a linear regression cannot do. Finally, I changed the layer size, being the number of hidden layers inside the MLP algorithm changing and deciding upon the number of hidden nodes in the network. 

Originally, I kept the training iterations to 1000, the Activation Function to ReLU, and the hidden layer size to 100. When I change these variables below, I keep the other two variables in their original state. 



| Configuration | Mean Accuracy | Standard Deviation |
|-----------------------------------|---------------|---------------------|
| Base State | 0.94 | 0.02 |
| Training Iteration -> 200 | 0.88 | 0.06 |
| Training Iteration -> 500 | 0.93 | 0.04 |
| Activation Function -> logistic | 0.92 | 0.03 |
| Activation Function -> identity | 0.94 | 0.02 |
| Hidden Layer Size -> 1 | 0.71 | 0.24 |
| Hidden Layer Size -> 500 | 0.95 | 0.04 |




When training the Neural Network, I changed up three variables inside the MLPClassifier algorithm that ultimalely trains the neural network. These three variables are the following. The training iterations, deciding how many times should this neural network be ran and the weights inside of it changed. 

Above are the results. 
Changing the activation function ultimately didn't change up the mean accuracy that much, with every different activation function appearing in all the others bounds within one standard deviation. 

Changing the Hidden Layer Size, on the other hand, had a bigger change with a single layer size having the lowest accuracy so far, and 500 layers having the highest mean accuracy layer. This can mean that while more hidden layers are better, the returns start to even out, with far less progress made from 100->500 layers than from 1->100 layers.

Changing the Training Iterations, also works as expected, with a higher accuracy resulting when more iterations are used, due to the algortihm being able to train itself and minimize the loss function more. The lowest training interation, has a smaller accuracy of 88 with the highest standard deviation, as the funcion has not been fully optimized, and with 500, a higher accuracy with a smaller deviation, as a result of the weights becoming more fitted to the equation.

Overall in this dataset, I think the activation function does not make that much of a difference, given it is one of the three measured


.
.
.
.
.
.
.
.
.
.
.

.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.

.
.
.
.

.

.

.

.

.

.

.

.


.

.

.

.

.


.


.


.





































